ui/uihandler.go

```diff
@@ -203,269 +138,139 @@ Use with caution, will interrupt/end this server: <input type="submit" name="exi
-					res.RequestedQPS, res.ActualQPS, res.NumThreads, res.RequestedDuration, fhttp.RoundDuration(res.ActualDuration),
-					percStr)))
-				// nolint: errcheck
-				w.Write([]byte(`],
-    },
-    elements: {
-     line: {
-      tension: 0, // disables bezier curves
-     }
-    },
-    scales: {
-      xAxes: [
-       linearXAxe
-      ],
-      yAxes: [{
-       id: 'P',
-       position: 'right',
-       ticks: {
-        beginAtZero: true,
-       },
-       scaleLabel : {
-        display: true,
-        labelString: '%'
-       }
-       },
-      linearYAxe
-      ]
-    }
-  }
-});
-function updateChart() {
-	var form = document.getElementById('updtForm')
-	var formMin = form.xmin.value.trim()
-	var formMax = form.xmax.value.trim()
-	var scales = chart.config.options.scales
-	var newAxis
-	var newXMin = parseFloat(formMin)
-	if (form.xlog.checked) {
-		newXAxis = logXAxe
-		//if (formMin == "0") {
-		//	newXMin = dataH[0].x // log doesn't like 0 xaxis
-		//}
-	} else {
-		newXAxis = linearXAxe
+			log.Errf("Error adding custom headers: %v", err)
+		}
 	}
-	if (form.ylog.checked) {
-		chart.config.options.scales = {xAxes: [newXAxis], yAxes: [scales.yAxes[0], logYAxe]}
-	} else {
-		chart.config.options.scales = {xAxes: [newXAxis], yAxes: [scales.yAxes[0], linearYAxe]}
+	out := io.Writer(w)
+	if JSONOnly {
+		out = os.Stderr
 	}
-	chart.update()
-	var newNewXAxis = chart.config.options.scales.xAxes[0]
-	if (formMin != "") {
-		newNewXAxis.ticks.min = newXMin
-	} else {
-		delete newNewXAxis.ticks.min
+	ro := periodic.RunnerOptions{
+		QPS:         qps,
+		Duration:    dur,
+		Out:         out,
+		NumThreads:  c,
+		Resolution:  resolution,
+		Percentiles: percList,
+		Labels:      labels,
 	}
-	if (formMax != "" && formMax != "max") {
-		newNewXAxis.ticks.max = parseFloat(formMax)
-	} else {
-		delete newNewXAxis.ticks.max
+	o := fhttp.HTTPRunnerOptions{
+		RunnerOptions: ro,
+		URL:           url,
 	}
-	chart.update()
+	res, err := fhttp.RunHTTPTest(&o)
+	if err != nil {
+		w.Write([]byte(fmt.Sprintf("Aborting because %v\n", err))) // nolint: errcheck
+		return
+	}
+	if JSONOnly {
+		w.Header().Set("Content-Type", "application/json")
+		j, err := json.MarshalIndent(res, "", "  ")
+		if err != nil {
+			log.Fatalf("Unable to json serialize result: %v", err)
+		}
+		_, err = w.Write(j)
+		if err != nil {
+			log.Errf("Unable to write json output for %v: %v", r.RemoteAddr, err)
+		}
+		return
+	}
+	// nolint: errcheck
+	w.Write([]byte(fmt.Sprintf("All done %d calls %.3f ms avg, %.1f qps\n</pre>\n<script>\n",
+		res.DurationHistogram.Count,
+		1000.*res.DurationHistogram.Avg,
+		res.ActualQPS)))
+	ResultToJsData(w, res)
+	ResultToChart(w, res)
+	w.Write([]byte("</script></body></html>\n"))
 }
-</script>
-</body></html>`))
-			}
+
+// ResultToJsData converts a result object to chart data arrays.
+func ResultToJsData(w io.Writer, res *fhttp.HTTPRunnerResults) {
```

Comment:
this looks hacky and hard to maintain. Could we just get a json string and do JSON.parse()? - such that we use go code to generate as less JS code as possible.

Reply:
It's turning histogramdata -> chart.js compatible array of data points so I don't think it's that hacky or hard to maintain, _but_
yes, moving toward processing from JS the actual Jason result is the next step, to move the code to JS but for now it's still incrementally much much better than the previous version, no ?